src/decoder/lattice-incremental-decoder.cc

```diff
@@ -0,0 +1,1617 @@
+                                                          int32 last_frame_of_chunk,
+                                                          CompactLattice *olat) {
+  using namespace fst;
+  bool not_first_chunk = last_get_lattice_frame_ != 0;
+  bool ret = true;
+
+  // last_get_lattice_frame_ is used to record the first frame of the chunk
+  // last time we obtain from calling this function. If it reaches
+  // last_frame_of_chunk
+  // we cannot generate any more chunk
+  if (last_get_lattice_frame_ < last_frame_of_chunk) {
+    Lattice raw_fst;
+    // step 1: Get lattice chunk with initial and final states
+    // In this function, we do not create the initial state in
+    // the first chunk, and we do not create the final state in the last chunk
+    if (!GetIncrementalRawLattice(&raw_fst, use_final_probs, last_get_lattice_frame_,
+                                  last_frame_of_chunk, not_first_chunk,
+                                  !decoding_finalized_))
+      KALDI_ERR << "Unexpected problem when getting lattice";
+    // step 2-3
+    ret = determinizer_.ProcessChunk(raw_fst, last_get_lattice_frame_,
+                                     last_frame_of_chunk);
+    last_get_lattice_frame_ = last_frame_of_chunk;
+  } else if (last_get_lattice_frame_ > last_frame_of_chunk)
+    KALDI_WARN << "Call GetLattice up to frame: " << last_frame_of_chunk
+               << " while the determinizer_ has already done up to frame: "
+               << last_get_lattice_frame_;
+
+  // step 4
+  if (decoding_finalized_) ret &= determinizer_.Finalize();
+  if (olat) {
+    *olat = determinizer_.GetDeterminizedLattice();
+    ret &= (olat->NumStates() > 0);
+  }
+  if (!ret) {
+    KALDI_WARN << "Last chunk processing failed."
+               << " We will retry from frame 0.";
+    // Reset determinizer_ and re-determinize from
+    // frame 0 to last_frame_of_chunk
+    last_get_lattice_frame_ = 0;
+    determinizer_.Init();
+  }
+
+  return ret;
+}
+
+template <typename FST, typename Token>
+bool LatticeIncrementalDecoderTpl<FST, Token>::GetIncrementalRawLattice(
+    Lattice *ofst, bool use_final_probs, int32 frame_begin, int32 frame_end,
+    bool create_initial_state, bool create_final_state) {
+  typedef LatticeArc Arc;
+  typedef Arc::StateId StateId;
+  typedef Arc::Weight Weight;
+  typedef Arc::Label Label;
+
+  if (decoding_finalized_ && !use_final_probs)
+    KALDI_ERR << "You cannot call FinalizeDecoding() and then call "
+              << "GetIncrementalRawLattice() with use_final_probs == false";
+
+  unordered_map<Token *, BaseFloat> final_costs_local;
+
+  const unordered_map<Token *, BaseFloat> &final_costs =
+      (decoding_finalized_ ? final_costs_ : final_costs_local);
+  if (!decoding_finalized_ && use_final_probs)
+    ComputeFinalCosts(&final_costs_local, NULL, NULL);
+
+  ofst->DeleteStates();
+  unordered_multimap<int, StateId>
+      token_label2last_state_map; // for GetInitialRawLattice
```

Comment:
this variable name, and also the tok_map/tmp_vec later -- I think I was told that it's better to not have words like map/vec in variable names. Not sure if it's really a rule, but here token_label2last_state_map, _map is redundant; the tok_map though isn't descriptive enough; maybe change it to token2state_id?

Reply:
I change them to token_label2last_state